compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -452,6 +452,11 @@ class TreeChecker extends Phase with SymTransformer {
       tree1
     }
 
+    override def typedWhileDo(tree: untpd.WhileDo)(implicit ctx: Context): Tree = {
+      assert((tree.cond ne EmptyTree) || ctx.phase.refChecked, i"invalid empty condition in while at $tree")
```

Comment:
Why `ctx.phase.refChecked`? Why not an equivalent method for `TailRec`?

Reply:
Because it existed, and the only requirement is that an infinite loop must not be pickled. It doesn't have anything to do with tailrec per se.